File: AVLTree/Java/Node.java
/**
 * Class representing a single Node of an AVL tree.
 * A Node contains a value that is also the insertion
 * key and references to its children and parent Nodes.
 *
 * @author Thomas Lang
 * @version 2014/09/02
 */
public class Node<T extends Comparable<T>> {

    private T value;
    private Node<T> left, right, parent;

    /**
     * Creates a new Node containing
     * the value <code>value</code>.
     *
     * @param value The value saved in the Node.
     */
    public Node( T value ){
        this.value = value;
	    left       = null;
    	right      = null;
	    parent     = null;
    }

    /**
     * Returns the saved value.
     *
     * @return The saved value in the Node.
     */
    public T getValue(){ return value; }

    /**
     * Returns the balance of the Node.
     *
     * @return The balance of the Node.
     */
    public int getBalance(){
        if(( left == null ) && ( right == null ))
	        return 0;
	    return ( height( right ) - height( left ) );
    }

    /**
     * Helper function for calculating the height 
     * of the Node <code>node</code>.
     *
     * @param  node The node to determine its height.
     * @return The height of <code>node</code>.
     */
    private int height( Node<T> node ){
	    if( node == null )
	        return 0;
	    return 1 + Math.max( height( node.getLeft() ), height( node.getRight() ));
    }

    /**
     * Returns the reference to the left child.
     *
     * @return The reference to the left child.
     */
    public Node<T> getLeft(){ return left; }

    /**
     * Returns the reference to the right child.
     *
     * @return The reference to the right child.
     */
    public Node<T> getRight(){ return right; }

    /**
     * Returns the reference to the parent Node.
     *
     * @return The reference to the parent Node.
     */
    public Node<T> getParent(){ return parent; }

    /**
     * Sets the value of the Node.
     *
     * @param value The new value to store.
     */
    public void setValue( T value ){
        this.value = value;
    }

    /**
     * Sets the left reference to <code>node</code>.
     *
     * @param node The new left reference.
     */
    public void setLeft( Node<T> node ){
        left = node;
    }

    /**
     * Sets the right reference to <code>node</code>.
     *
     * @param node The new right reference.
     */
    public void setRight( Node<T> node ){
        right = node;
    }

    /**
     * Sets the parent reference to <code>node</code>.
     *
     * @param node The new parent reference.
     */
    public void setParent( Node<T> node ){
        parent = node;
    }

    /**
     * Returns an appropriate String representation
     * of the calling Node.
     *
     * @return A String representation of the Node.
     */
    public String toString(){
        return ("{" + value + "}");
    }
}

File: AVLTree/C/node.c
/* Copyright (c) 2015, Thomas Lang. All rights reserved.
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>

#include "node.h"

/*
 * Creates a new Node with the passed {@code value}.
 *
 * @param value
 *          The value hold by the newly created node.
 */
Node* newNode(int value) {
    Node *node   = (Node*) malloc(sizeof(Node));
    node->value  = value;
    node->left   = NULL;
    node->right  = NULL;
    node->parent = NULL;
    return node;
}

/*
 * Returns the height of this node in the surrounding tree.
 *
 * @return Returns the height of this node in the surrounding tree.
 */
static int __getHeight(Node *node) {
    assert (node != NULL);

    if (!(node->left) && !(node->right)) {
        return 1;
    }

    int rh = !node->left  ? 0 : __getHeight(node->left);
    int lh = !node->right ? 0 : __getHeight(node->right);
    return 1 + MAX(lh, rh);
}

/*
 * Returns the <em>balance</em> of this node, which is simply the
 * height of its right sub tree minored by the height of the left
 * sub tree.<p>
 * This value is essential for tree rotations that will stabilize the
 * tree after every operation if necessary.
 *
 * @return Returns the balance of this node.
 * @see AVLTree#rotate(Node, boolean)
 */
int getBalance(Node *node) {
    assert (node != NULL);

    if (!(node->left) && !(node->right)) {
        return 0;
    }

    int lbal = !node->left  ? 0 : __getHeight(node->left);
    int rbal = !node->right ? 0 : __getHeight(node->right);
    return rbal - lbal;
}

/*
 * Prints out this node including all children.
 *
 * @param node
 *          The node to print.
 */
void printNode(Node *node) {
    assert (node != NULL);

    if (node->left) {
        printf("(");
        printNode(node->left);
        printf(")");
    }

    printf("[%d]", node->value);

    if (node->right) {
        printf("(");
        printNode(node->right);
        printf(")");
    }
}
